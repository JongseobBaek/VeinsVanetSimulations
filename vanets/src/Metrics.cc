#include "Metrics.h"

Define_Module(Metrics);

void Metrics::initialize(){

    currentSimulationTime = 0;
    packetsDeliveredToMetrics = 0;
    throughputMetric = 0;
    throughputSignal = registerSignal("throughputSignal");

    cMessage *timerMessage = new cPacket();
    timerMessage->setName("timer");

    cMessage *dataMessage = new cPacket();
    dataMessage->setName("data");

    scheduleAt(simTime(),timerMessage);
    scheduleAt(simTime(),dataMessage);
    WATCH(packetsDeliveredToMetrics);
    WATCH(throughputMetric);

}

double Metrics::computeThroughput(double currentPacketsReceived,double currentSimulationTime) {

    double result = 0;
    simtime_t testSimTime=simTime();

    std::cout<< "Test Time double: "<<testSimTime.dbl()<<std::endl;
    std::cout<< "Simulation Time: "<<currentSimulationTime<<std::endl;
    std::cout<< "Test Time simetime_t type:"<<testSimTime<<std::endl;

    if (currentSimulationTime>0){
        std::cout<<"currentPacketsReceived type: "<<typeid(currentPacketsReceived).name()<<'\n';
        std::cout<<"currentSimulationTime type: "<<typeid(currentSimulationTime).name()<<'\n';
        result = currentPacketsReceived/currentSimulationTime;
    }
    return result;
}

simtime_t Metrics::getCurrentTime(){
    return currentSimulationTime;
}

void Metrics::handleMessage(cMessage *msg)
{
     if(msg->isSelfMessage()){
         currentSimulationTime = simTime();
         scheduleAt(simTime()+1,msg);
     }
     std::cout<<msg<<" "<<currentSimulationTime<<std::endl;

     if(msg->isName("data")){
         packetsDeliveredToMetrics = updateNumberOfPacketsReceived(packetsDeliveredToMetrics);
         double currentSimulationTimeDouble = currentSimulationTime.dbl();
         throughputMetric = computeThroughput(packetsDeliveredToMetrics, currentSimulationTimeDouble);
         std::cout<<"Current Throughput:"<<throughputMetric<<std::endl;
         emit(throughputSignal,throughputMetric);
     }


}

double Metrics::updateNumberOfPacketsReceived(double packetsDeliveredToMetrics){
     return packetsDeliveredToMetrics = packetsDeliveredToMetrics +1;
}

void Metrics::finish(){
    recordScalar("packetsDeliveredToMetrics", packetsDeliveredToMetrics);
    recordScalar("throughputMetric", throughputMetric);
}
