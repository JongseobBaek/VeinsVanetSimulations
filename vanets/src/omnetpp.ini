[General]
cmdenv-express-mode = true
cmdenv-autoflush = true
cmdenv-status-frequency = 10000000s

#tkenv-default-config = debug
#tkenv-default-run = 1
tkenv-image-path = bitmaps
ned-path = .

network = RSUScenario

##########################################################
#            Simulation parameters                       #
########################mac##################################
debug-on-errors = true
print-undisposed = false
record-eventlog = true

#6000s
sim-time-limit = 200s

#Extra parameters for testing signals
check-signals=true

#Testing parameters
**.scalar-recording = true
**.vector-recording = true
**.throughputSignal.scalar-recording = true

**.debug = true
**.coreDebug = false

*.playgroundSizeX = 2500m
*.playgroundSizeY = 2500m
*.playgroundSizeZ = 50m

##########################################################
# Annotation parameters                                  #
##########################################################
*.annotations.draw = false

##########################################################
# Obstacle parameters                                    #
##########################################################
*.obstacles.debug = false

##########################################################
#WorldUtilityParameters									 #
##########################################################
*.world.useTorus = false
*.world.use2D = false

##########################################################
#TraCIScenarioManager Parameters						 #
##########################################################
*.manager.updateInterval = 0.1s
*.manager.host = "localhost"
*.manager.port = 9999
*.manager.moduleType = "vanets.src.CarWithMetrics"
*.manager.moduleName = "node"
*.manager.moduleDisplayString = ""
*.manager.autoShutdown = true
*.manager.margin = 25
*.manager.numVehicles = 5
*.manager.launchConfig = xmldoc("erlangen.launchd.xml")

##########################################################
# RSU SETTINGS											 #
##########################################################
*.rsu[0].mobility.x = 2000
*.rsu[0].mobility.y = 2000
*.rsu[0].mobility.z = 3
*.rsu[*].applType = "TraCIDemoRSU11p"
*.rsu[*].appl.debug = false
*.rsu[*].appl.headerLength = 256 bit
*.rsu[*].appl.sendBeacons = false
*.rsu[*].appl.dataOnSch = false
*.rsu[*].appl.sendData = true
*.rsu[*].appl.beaconInterval = 1s
*.rsu[*].appl.beaconPriority = 3
*.rsu[*].appl.dataPriority = 2
*.rsu[*].appl.maxOffset = 0.005s

*.metricsApplication.mobility.x = 2000
*.metricsApplication.mobility.y = 2000
*.metricsApplication.mobility.z = 3
*.metricsApplication.applType = "TraCIDemoRSU11p"
*.metricsApplication.appl.debug = false
*.metricsApplication.appl.headerLength = 256 bit
*.metricsApplication.appl.sendBeacons = false
*.metricsApplication.appl.dataOnSch = false
*.metricsApplication.appl.sendData = true
*.metricsApplication.appl.beaconInterval = 1s
*.metricsApplication.appl.beaconPriority = 3
*.metricsApplication.appl.dataPriority = 2
*.metricsApplication.appl.maxOffset = 0.005s


##########################################################
#            11p specific parameters                     #
#                                                        #
#                    NIC-Settings                        #
##########################################################
*.connectionManager.pMax = 20mW
*.connectionManager.sat = -89dBm
*.connectionManager.alpha = 2.0
*.connectionManager.carrierFrequency = 5.890e9 Hz
*.connectionManager.sendDirect = true

*.**.nic.mac1609_4.useServiceChannel = false
*.**.nic.mac1609_4.scalar-recording = true
 
*.**.nic.mac1609_4.txPower = 20mW
*.**.nic.mac1609_4.bitrate = 6Mbps
*.**.nic.phy80211p.sensitivity = -89dBm
*.**.nic.phy80211p.maxTXPower = 100mW
*.**.nic.mac1609_4.queueSize = ${1,2,5}
*.**.nic.mac1609_4.vector-recording = true

*.**.nic.phy80211p.useThermalNoise = true
*.**.nic.phy80211p.thermalNoise = -110dBm
*.**.nic.phy80211p.decider = xmldoc("config.xml")
*.**.nic.phy80211p.analogueModels = xmldoc("config.xml")
*.**.nic.phy80211p.usePropagationDelay = true

##########################################################
##########################################################
*.node[*].applType = "TraCIDemo11p"
*.node[*].appl.debug = false
*.node[*].appl.headerLength = 256 bit
*.node[*].appl.sendBeacons = false
*.node[*].appl.dataOnSch = false
*.node[*].appl.sendData = true
*.node[*].appl.beaconInterval = 1s
*.node[*].appl.beaconPriority = 3
*.node[*].appl.dataPriority = 2
*.node[*].appl.maxOffset = 0.005s



##########################################################
#                      Mobility                          #
##########################################################
*.node[*].veinsmobilityType = "org.car2x.veins.modules.mobility.traci.TraCIMobility"
*.node[*].mobilityType = "TraCIMobility"
*.node[*].mobilityType.debug = true
*.node[*].veinsmobilityType.debug = true
*.node[*].veinsmobility.x = 0
*.node[*].veinsmobility.y = 0
*.node[*].veinsmobility.z = 1.895
*.node[*0].veinsmobility.accidentCount = 1
*.node[*0].veinsmobility.accidentStart = 10s #75s
*.node[*0].veinsmobility.accidentDuration = 30s

[Config nodebug]
description = "default settings"
**.debug = false
**.coreDebug = false
*.annotations.draw = false

[Config debug]
description = "(very slow!) draw and print additional debug information"
**.debug = true
**.coreDebug = true
*.annotations.draw = true

[Config B_400Bytes_P_1000Bytes_Q_5]
description = "BeaconLength 400Bytes, ServicePacketLength 1000Bytes, QueueSize 5"
**.debug = false
**.coreDebug = false
*.annotations.draw = false
*.rsu[*].appl.beaconLengthBits = 3200bit
*.node[*].appl.beaconLengthBits = 3200bit
*.rsu[*].appl.dataLengthBits = 8000bit
*.node[*].appl.dataLengthBits = 8000bit
*.rsu[*].appl.dataOnSch = true
*.node[*].appl.dataOnSch = true
*.**.nic.mac1609_4.queueSize = 5
*.**.nic.mac1609_4.useServiceChannel = true
*.manager.launchConfig = xmldoc("singlechannel/LowDensityScenario.launchd.xml")

