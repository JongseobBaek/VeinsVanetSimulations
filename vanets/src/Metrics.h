//Reference to Andres Vargas class sink.h in inet 

#ifndef __INET_INET_METRICS_H
#define __INET_INET_METRICS_H


#include <omnetpp.h>
#include <FindModule.h>
/**
 * A module that just deletes every packet it receives, and collects
 * basic statistics (packet count, bit count, packet rate, bit rate).
 */
class  Metrics : public cSimpleModule
{
public:
    virtual ~Metrics(){}
    double throughputMetric;

  protected:
    simtime_t currentSimulationTime;
    double packetsDeliveredToMetrics;

    simsignal_t throughputSignal;

  protected:
    virtual void initialize();
    virtual void handleMessage(cMessage *msg);
    virtual void finish();
    virtual double updateNumberOfPacketsReceived(double packetsDeliveredToMetrics);
    virtual simtime_t getCurrentTime();
  public:
    virtual double computeThroughput(double packetsDeliveredToMetrics, double currentSimulationTime);
    double getThroughputMetric();
    double getThroughputSignal();
};

#endif /* INET_INET_METRICS_H_ */


