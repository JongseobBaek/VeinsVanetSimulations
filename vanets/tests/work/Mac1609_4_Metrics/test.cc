
#include <omnetpp.h>

#include <Mac16094MetricsTest.h>
#include <../../../src/Metrics.h>



namespace Mac1609_4_Metrics {



class Test : public cSimpleModule
{
    public:
        Test() : cSimpleModule(32768) {}
        virtual void activity();
};

Define_Module(Test);

void Test::activity()
{

Metrics *metrics = new Metrics();
Mac16094MetricsTest *macMetrics = new Mac16094MetricsTest();
double simulationTime = 0;
double currentPacketsReceived = 0;

//when simulation time is zero get zero
simulationTime = 0;
currentPacketsReceived = 0;
macMetrics->testComputeThroughputPacketsPerSecond(metrics, currentPacketsReceived, simulationTime);

//when currentPackets is zero
currentPacketsReceived = 0;
simulationTime =5;
macMetrics->testComputeThroughputPacketsPerSecond(metrics, currentPacketsReceived, simulationTime);

currentPacketsReceived = 5;
simulationTime = 10;
macMetrics->testComputeThroughputPacketsPerSecond(metrics, currentPacketsReceived, simulationTime);

cPacket *pkt = new cPacket();
cMessage *message = pkt; //Valid cPacket descendant of cMessage
macMetrics->testHandleLowerMsgCallsThroughputWhenReceivingBroadcast(message);

//when currentMgbis is zero
double currentMgbits = 0;
cPacket *packet = dynamic_cast<cPacket*>(message);
packet->setBitLength(1);
simulationTime = 10;
macMetrics->testThroughputMegabitsPerSecond(metrics, 1, currentMgbits, simulationTime);

//when currentMbgits is not zero
currentMgbits = 2;
packet->setBitLength(4000000);
simulationTime = 10;
macMetrics->testThroughputMegabitsPerSecond(metrics, 4000000, currentMgbits, simulationTime);

//when packetLength is zero
currentMgbits = 2;
packet->setBitLength(0);
simulationTime = 10;
macMetrics->testThroughputMegabitsPerSecond(metrics, 0, currentMgbits, simulationTime);

//when currentMgbits is a really big digit number
currentMgbits = 2000000000000000000000000000;
packet->setBitLength(2000000000000000000);
simulationTime = 10;
macMetrics->testThroughputMegabitsPerSecond(metrics, 2000000000000000000, currentMgbits, simulationTime);

macMetrics->testGetAvgThroughput("B_400Bytes_P_1000Bytes_Q_5-0.sca");


}

}; //namespace
