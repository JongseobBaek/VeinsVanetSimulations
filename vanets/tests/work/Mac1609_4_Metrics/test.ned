package vanets.tests.work.Mac1609_4_Metrics;

simple Test{
	@isNetwork(true);
}

