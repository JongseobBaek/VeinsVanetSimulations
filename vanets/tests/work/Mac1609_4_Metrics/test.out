OMNeT++ Discrete Event Simulation  (C) 1992-2014 Andras Varga, OpenSim Ltd.
Version: 4.6, build: 141202-f785492, edition: Academic Public License -- NOT FOR COMMERCIAL USE
See the license for distribution terms and warranty disclaimer
Setting up Cmdenv...
Loading NED files from .: 2

Preparing for running configuration General, run #0...
Scenario: $repetition=0
Assigned runID=General-0-20160808-14:05:49-8824
Setting up network `Test'...
Initializing...
Initializing module Test, stage 0

Running simulation...
** Event #1  T=0   Test (Test, id=1)
throughputPps: 0.0000000000000000
throughputPps: 0.0000000000000000
throughputPps: 0.5000000000000000
throughputMbps: 0.0000001000000000
throughputMbps: 0.6000000000000000
throughputMbps: 0.2000000000000000
throughputMbps: 458494841882057920.0000000000000000

<!> No more events -- simulation ended at event #2, t=0.


Calling finish() at end of Run #0...
undisposed object: (cPacket) .Test. -- check module destructor
undisposed object: (Metrics)  -- check module destructor
undisposed object: (Mac16094MetricsTest)  -- check module destructor

End.
